_posts/2020-05-04-this-week-in-glean.md

```diff
@@ -0,0 +1,266 @@
+
+[jna]: https://github.com/java-native-access/jna
+
+First some imports and replicating the variants our tag takes.
+
+```java
+import com.sun.jna.Structure
+import com.sun.jna.Pointer
+import com.sun.jna.Union
+
+enum class TaskTag {
+    Upload,
+    Wait,
+    Done
+}
+```
+
+Next is the body of our `Upload` variant. It's a structure with two pointers to strings.
+Kotlin requires some annotations to specify the order of fields in memory.
+We also  inherit from `Structure`, a class provided by JNA, that will take care of reading from memory and making the data accessible in Kotlin.
+
+```java
+@Structure.FieldOrder("tag", "id", "url")
+class UploadBody(
+    @JvmField val tag: Byte = TaskTag.Done.ordinal.toByte(),
+    @JvmField val id: Pointer? = null,
+    @JvmField val url: Pointer? = null,
+) : Structure() { }
+```
+
+And at last we define our union.
+We don't need a field order, it's a union afterall, only one of the fields is valid at a time.
+
+```java
+open class FfiTask(
+    @JvmField var tag: Byte = TaskTag.Done.ordinal.toByte(),
+    @JvmField var upload: UploadBody = UploadBody()
+) : Union() {
+    class ByValue : FfiTask(), Structure.ByValue
+
+    fun toTask(): Task {
+        this.readField("tag")
+        return when (this.tag.toInt()) {
+            TaskTag.Upload.ordinal -> {
+                this.readField("upload")
+                val request = Request(
+                    this.upload.id.getRustString(),
+                    this.upload.url.getRustString()
+                )
+                Task.Upload(request)
+            }
+            TaskTag.Wait.ordinal -> Task.Wait
+            else -> Task.Done
+        }
+    }
+}
+```
+
+This also defines the conversion to a new Kotlin type that eases usage on the Kotlin side.
+`getRustString` is a [small helper](https://github.com/mozilla/glean/blob/da1e014e378660f243b92ce6012b7b29188d72d8/glean-core/android/src/main/java/mozilla/telemetry/glean/rust/LibGleanFFI.kt#L40-L42)
+to copy the null-terminated C-like string to a Kotlin string.
+
+The FFI function on the Kotlin side is defined as:
+
+```java
+fun glean_get_next_task(): FfiTask.ByValue
+```
+
+The types we convert to and then work with in Kotlin are small classes around the data.
+If there's no attached data it's an object.
+
+```java
+class Request(
+    val id: String,
+    val url: String,
+) { }
+
+sealed class Task {
+    class Upload(val request: Request) : Task()
+
+    object Wait : Task()
+
+    object Done : Task()
+}
+```
+
+As the final piece of this code on the Kotlin side we can now fetch new tasks, convert it to more convenient Kotlin objects and work with them:
+
+```java
+val incomingTask = LibGleanFFI.INSTANCE.glean_get_next_task()
+when (val action = incomingTask.toTask()) {
+    is Task.Upload -> upload(action.request.id, aciton.request.url)
+    Task.Wait -> return Result.retry()
+    Task.Done -> return Result.success()
+}
+```
+
+## What's next?
+
+Currently this piece of code is under testing.
```

Comment:
The upload parts are under testing, but the more general "passing data through FFI" part has been around for a long while. I'd recommend clarifying that a bit.